Gems/Atom/RPI/Code/Source/RPI.Public/Pass/CopyPass.cpp

```diff
@@ -84,6 +92,70 @@ namespace AZ
                 "CopyPass must have exactly 2 bindings: 1 input and 1 output. %s has %d bindings.",
                 GetPathName().GetCStr(), m_attachmentBindings.size());
 
+            bool sameDevice = (m_data.m_sourceDeviceIndex == -1 && m_data.m_destinationDeviceIndex == -1) ||
+                m_data.m_sourceDeviceIndex == m_data.m_destinationDeviceIndex;
+            AZ_Assert(
+                sameDevice || (m_data.m_sourceDeviceIndex != -1 && m_data.m_destinationDeviceIndex != -1),
+                "CopyPass: Either source and destination device indices must be invalid, or both must be valid");
+
+            m_copyMode = sameDevice ? CopyMode::SameDevice : CopyMode::DifferentDevicesIntermediateHost;
+
+            if (m_copyMode == CopyMode::SameDevice)
+            {
+                m_copyScopeProducerSameDevice = AZStd::make_shared<RHI::ScopeProducerFunctionNoData>(
+                    RHI::ScopeId{ GetPathName() },
+                    AZStd::bind(&CopyPass::SetupFrameGraphDependenciesSameDevice, this, AZStd::placeholders::_1),
+                    AZStd::bind(&CopyPass::CompileResourcesSameDevice, this, AZStd::placeholders::_1),
+                    AZStd::bind(&CopyPass::BuildCommandListInternalSameDevice, this, AZStd::placeholders::_1),
+                    m_hardwareQueueClass);
+            }
+            else if (m_copyMode == CopyMode::DifferentDevicesIntermediateHost)
+            {
+                [[maybe_unused]] auto* device1 =
+                    RHI::RHISystemInterface::Get()->GetDevice(m_data.m_sourceDeviceIndex != RHI::MultiDevice::InvalidDeviceIndex ? m_data.m_sourceDeviceIndex : RHI::MultiDevice::DefaultDeviceIndex);
+                AZ_Assert(
+                    device1->GetFeatures().m_signalFenceFromCPU,
+                    "CopyPass: Device to device copy is only possible if all devices support signalling fences from the CPU");
+                [[maybe_unused]] auto* device2 =
+                    RHI::RHISystemInterface::Get()->GetDevice(m_data.m_destinationDeviceIndex != RHI::MultiDevice::InvalidDeviceIndex ? m_data.m_destinationDeviceIndex : RHI::MultiDevice::DefaultDeviceIndex);
+                AZ_Assert(
+                    device2->GetFeatures().m_signalFenceFromCPU,
+                    "CopyPass: Device to device copy is only possible if all devices support signalling fences from the CPU");
+
+                for (auto& fence : m_device1SignalFence)
+                {
+                    fence = new RHI::MultiDeviceFence();
+                    AZ_Assert(fence != nullptr, "CopyPass failed to create a fence");
+                    [[maybe_unused]] RHI::ResultCode result = fence->Init(RHI::MultiDevice::AllDevices, RHI::FenceState::Signaled);
+                    AZ_Assert(result == RHI::ResultCode::Success, "CopyPass failed to init fence");
+                }
+
+                for (auto& fence : m_device2WaitFence)
```

Comment:
Recommend adding a quick comment about initializing fences about max frames and how they will be used.